### PR TITLE
Check for presence of null case in MissingWhenCase rule

### PR DESCRIPTION
**Pull request by sowmyav24**
Fixes: [#3189 ](https://github.com/detekt/detekt/issues/3189)

- `WhenChecker` does not return `null` while fetching missing cases.
- Used the method `containsNullCase` to verify if null was defined by user.

### CODE REVIEW

**Review by BraisGabin — DISMISSED** [review]

**Comment by t-kameyama** [comment]
False positive:
```kotlin
enum class Color {
    RED,
    GREEN,
    BLUE
}
fun whenOnEnumFail(c: Color?) {
    if (c == null) return
    when (c) {
        Color.RED -> {}
        Color.BLUE -> {}
        Color.GREEN -> {}
    }
}
```

**Comment by sowmyav24** [comment]
> False positive:
> 
> ```kotlin
> enum class Color {
>     RED,
>     GREEN,
>     BLUE
> }
> fun whenOnEnumFail(c: Color?) {
>     if (c == null) return
>     when (c) {
>         Color.RED -> {}
>         Color.BLUE -> {}
>         Color.GREEN -> {}
>     }
> }
> ```

@BraisGabin Do you think we can check if null check has been done prior to `when` block execution ?

File: detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCaseSpec.kt [hunk 1]
```diff
@@ -38,6 +38,49 @@ object MissingWhenCaseSpec : Spek({
                 assertThat(actual.first().message).isEqualTo("When expression is missing cases: RED. Either add missing cases or a default `else` case.")
             }
 
+            it("reports when `when` expression used as statement and not all cases including null are not covered") {
```
Comment (schalkms) [inline detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCaseSpec.kt:41]: Double negation:

```suggestion
            it("reports when `when` expression used as statement and not all cases including null are covered") {
```

File: detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCaseSpec.kt [hunk 2]
```diff
@@ -87,6 +130,49 @@ object MissingWhenCaseSpec : Spek({
...
+                        }
+                    }
+                """
+                val actual = subject.compileAndLintWithContext(env, code)
+                assertThat(actual).hasSize(1)
+                assertThat(actual.first().issue.id).isEqualTo("MissingWhenCase")
+                assertThat(actual.first().message).isEqualTo("When expression is missing cases: null. Either add missing cases or a default `else` case.")
+            }
+
+            it("reports when `when` expression used as statement and not all cases including null are not covered") {
```
Comment (schalkms) [inline detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCaseSpec.kt:155]: Double negation:

```suggestion
            it("reports when `when` expression used as statement and not all cases including null are covered") {
```

File: detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCaseSpec.kt (hunk 1 above)
Comment (sowmyav24) [reply detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCaseSpec.kt]: Done 👍

File: detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCaseSpec.kt (hunk 2 above)
Comment (sowmyav24) [reply detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/MissingWhenCaseSpec.kt]: Done 👍

**Comment by BraisGabin** [comment]
> > False positive:
> > ```kotlin
> > enum class Color {
> >     RED,
> >     GREEN,
> >     BLUE
> > }
> > fun whenOnEnumFail(c: Color?) {
> >     if (c == null) return
> >     when (c) {
> >         Color.RED -> {}
> >         Color.BLUE -> {}
> >         Color.GREEN -> {}
> >     }
> > }
> > ```
> 
> @BraisGabin Do you think we can check if null check has been done prior to `when` block execution ?

🤔 Can we know about the smart casting? This case is trivial and we could say "you should add the `null` inside the `when`". But I can imagine more complex code where you can't just add it in the `when`. So we should relay in the smart cast. If we can't do that I think that it's better to have a false-negative than a false-positive.

**Comment by schalkms** [comment]
> Can we know about the smart casting? This case is trivial and we could say "you should add the null inside the when". But I can imagine more complex code where you can't just add it in the when. So we should relay in the smart cast. If we can't do that I think that it's better to have a false-negative than a false-positive.

After reading the good comment from @t-kameyama I agree with @BraisGabin here. We should leave the null check out and stick with the false-negative. I think there are many more false-positives that look similar to the mentioned case, but are not the same. To detect those cases, you would need to use data-flow analysis to get it completely right. Currently, detekt does not have it and I'm unsure, whether detekt will have it in the near future. My tendency would be no.

**Comment by sowmyav24** [comment]
> To detect those cases, you would need to use data-flow analysis to get it completely right

I agree with this. To avoid `false positives`, I think we should go ahead and close this PR and the issue.

**Comment by schalkms** [comment]
Okay, I'll close it.
I'm sorry for you @sowmyav24 that your hard work on this PR isn't rewarded by a merge.

**Comment by schalkms** [comment]
Okay, I'll close it.
I'm sorry @sowmyav24 that your hard work on this PR isn't rewarded by a merge.

**Comment by sowmyav24** [comment]
> I'm sorry @sowmyav24 that your hard work on this PR isn't rewarded by a merge.

It is good we caught the issue before merging :)

**Comment by t-kameyama** [comment]
We can detect smart casting. 

See:
https://github.com/detekt/detekt/blob/b1269b58dc704c096ed402f91c9962942989a18a/detekt-rules-errorprone/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/NullableToStringCall.kt#L66-L78

https://github.com/detekt/detekt/blob/282c038c2e9fdc462b73952c65d1688b15fd35da/detekt-rules-errorprone/src/test/kotlin/io/gitlab/arturbosch/detekt/rules/bugs/NullableToStringCallSpec.kt#L73-L87

**Comment by BraisGabin** [comment]
@schalkms you said that smart cast was not enough for this issue, could you explain why? Otherwise we can implement this.

**Comment by schalkms** [comment]
I sent my comments. Please check the referenced issue. Thanks to @t-kameyama for mentioning the necessary data flow analysis code snippet in order to achieve this.

**Comment by BraisGabin** [comment]
@sowmyav24 if you want, you can continue with this PR. Thanks @t-kameyama !

**Comment by sowmyav24** [comment]
I will update the PR 👍

**Comment by sowmyav24** [comment]
In order to fetch the data flow information, we have to traverse and visit the complete block of function/expression. Currently `MissingWhenCase` takes into account only `visitWhenExpression`. Does anyone see any downside to that ?

**Comment by t-kameyama** [comment]
I think we should check `WhenChecker.getMissingCases(expression, bindingContext)`.

1. 
```
enum class Color {
    RED,
    GREEN,
    BLUE
}
fun whenOnEnumFail(c: Color?) {
    when (c) {
        Color.RED -> {}
        Color.BLUE -> {}
        Color.GREEN -> {}
    }
}
```
`WhenChecker.getMissingCases(expression, bindingContext)` is not empty.

2.
```
enum class Color {
    RED,
    GREEN,
    BLUE
}
fun whenOnEnumFail(c: Color?) {
    if (c == null) return
    when (c) {
        Color.RED -> {}
        Color.BLUE -> {}
        Color.GREEN -> {}
    }
}
```
`WhenChecker.getMissingCases(expression, bindingContext)` is empty.

**Comment by sowmyav24** [comment]
@t-kameyama Thank you for this input . I did not realise `getMissingCases` ensures it is taken care of.

**Comment by sowmyav24** [comment]
@BraisGabin @schalkms Please take a look at the latest commit changes. I have incorporated the suggestion given by @t-kameyama

**Review by BraisGabin — APPROVED** [review]

**Review by schalkms — APPROVED** [review]

**Pull request merged**